harmoni_detectors/harmoni_stt/nodes/deepspeech_service.py

```diff
@@ -0,0 +1,156 @@
+from std_msgs.msg import String
+import io
+import numpy as np
+import os
+
+
+class SpeechToTextService(HarmoniServiceManager):
+	"""
+	Speech to text service using DeepSpeech
+	"""
+
+	def __init__(self, name, param):
+		""" Initialization of variables and DeepSpeech parameters """
+		super().__init__(name)
+		self.model_file_path = param["model_file_path"]
+		self.scorer_path = param["scorer_path"]
+		self.lm_alpha = param["lm_alpha"]
+		self.lm_beta = param["lm_beta"]
+		self.beam_width = param["beam_width"]
+		self.t_wait = param["t_wait"]
+		self.subscriber_id = param["subscriber_id"]
+		# If False, stream transcription will end after DeepSpeech client determines its transcription as final (dependent on the t_wait param)
+		self.is_transcribe_continuously = param["is_transcribe_continuously"]
+
+		self.service_id = hf.get_child_id(self.name)
+
+		self.ds_client = DeepSpeechClient(
+			self.model_file_path,
+			self.scorer_path,
+			self.lm_alpha,
+			self.lm_beta,
+			self.beam_width,
+			self.t_wait
+		)
+
+		"""Setup publishers and subscribers"""
+		rospy.Subscriber("/audio/audio", AudioData, self.playing_sound_pause_callback)
+		rospy.Subscriber(
+			SensorNameSpace.microphone.value + self.subscriber_id,
+			AudioData,
+			self.sound_data_callback,
+		)
+		self.text_pub = rospy.Publisher(
+			DetectorNameSpace.stt.value + self.service_id, String, queue_size=10
+		)
+
+		self.state = State.INIT
+		return
+
+	def start(self, rate=""):
+		"""Start the DeepSpeech stream"""
+		rospy.loginfo("Start the %s service" % self.name)
+		if self.state == State.INIT or self.state == State.FAILED:
+			self.state = State.START
+			# Start the DeepSpeech client streaming inference state
+			self.ds_client.start_stream()
+		else:
+			self.state = State.START
+		return
+
+	def stop(self):
+		rospy.loginfo("Stop the %s service" % self.name)
+		try:
+			# Close the DeepSpeech streaming inference state
+			text = self.ds_client.finish_stream()
+			rospy.loginfo(f"Final text: {text}")
+			self.state = State.SUCCESS
+		except Exception:
+			self.state = State.FAILED
+		return
+
+	def pause(self):
+		rospy.loginfo("Pause the %s service" % self.name)
+		self.state = State.SUCCESS
+		return
+
+	def sound_data_callback(self, data):
+		"""Callback function subscribing to the microphone topic.
+		Passes audio data to DeepSpeech client.
+		"""
+		data = np.fromstring(data.data, np.uint8)
+		if self.state == State.START:
+			self.transcribe_stream_request(data)
+		# else:
+			# rospy.loginfo(f"Not Transcribing data because state is {self.state}")
+
+	def transcribe_stream_request(self, data):
+		stream = data
+		text = self.ds_client.process_audio(data)
+		rospy.loginfo(f"I heard: {text}")
+		if self.ds_client.is_final:
+		# Once the DeepSpeech client determines the text as final,
+		# the text will be published. 
+			rospy.loginfo(f"Final text: {text}")
+			self.text_pub.publish(text)
+			if not self.is_transcribe_continuously:
+				self.stop()
+		return
```

Comment:
I included the `is_transcribe_continuously` parameter to control whether or not the STT service automatically stops after the DeepSpeech client determines the text as final. Would it make more sense to remove this parameter and have this 'transcribe once' functionality in `request()` instead?

Reply:
Yeah, I like that. Having continuous transcription on the start() but then single instance on the request() would be perfect.